include/namedstruct/phantom_type.h

```diff
@@ -0,0 +1,245 @@
+/*
+The MIT License (MIT)
+
+Copyright (c) 2016 Yehonatan Ballas
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+#ifndef namedstruct_phantom_type_h
+#define namedstruct_phantom_type_h
+
+#include <limits>
+#include <type_traits>
+#include <utility>
+#include <optional>
+
+#ifdef NAMEDSTRUCT_DECLARE_MSGPACK
+#include <TransitUtils/msgpack_support.h>
+#endif
```

Comment:
`phantom_type.h` should have an optional dependency on `msgpack` only:
```suggestion
#ifdef NAMEDSTRUCT_MSGPACK_SUPPORT
#include <msgpack.hpp>
#endif
```